### ISSUE TITLE
RuntimeError: dictionary changed size during iteration in _handle_cancelled()

### ISSUE THREAD

**Issue by gsalgado** [ethereum/async-service#44]
This happened when one my service's daemon tasks terminated by raising an exception. The exception from the daemon task is lost and this was the only thing I got in the logs

```
13:26:13 DEBUG: starting lookup; initial neighbours: [<Node(0x15ac307a470b411745a6f10544ed54c0a14ad640b21f04f523e736e732bf709d4e28c2f06526ecabc03eed226b6d9bee8e433883cd20ab6cbd114bab77a8775d@52.176.7.10:30303)>, <Node(0x1836710284935d21d3c8ec8735ded61057b3b6700415c9a17643c0fbb79c0cc65acf478374efcba027e21549e845ad2a4bfc9a695579676ac5ffd974f4941fb7@58.243.201.60:42445)>, <Node(0x865a63255b3bb68023b6bffd5095118fcc13e79dcf014fe4e47e065c350c7cc72af2e53eff895f11ba1bbb6a2b33271c1116ee870f266618eadfc2e78aa7349c@52.176.100.77:30303)>]
Traceback (most recent call last):
  File "scripts/discovery.py", line 67, in <module>
    trio.run(main)
  File "/home/salgado/virtualenvs/trinity-trio/lib/python3.7/site-packages/trio/_core/_run.py", line 1804, in run
    raise runner.main_task_outcome.error
  File "scripts/discovery.py", line 63, in main
    await service.manager.wait_finished()
  File "/home/salgado/virtualenvs/trinity-trio/lib/python3.7/site-packages/async_generator/_util.py", line 42, in __aexit__
    await self._agen.asend(None)
  File "/home/salgado/src/async-service/async_service/trio.py", line 413, in background_trio_service
    await manager.stop()
  File "/home/salgado/virtualenvs/trinity-trio/lib/python3.7/site-packages/trio/_core/_run.py", line 730, in __aexit__
    raise combined_error_from_nursery
  File "/home/salgado/src/async-service/async_service/trio.py", line 140, in run
    system_nursery.cancel_scope.cancel()
  File "/home/salgado/virtualenvs/trinity-trio/lib/python3.7/site-packages/trio/_core/_run.py", line 730, in __aexit__
    raise combined_error_from_nursery
  File "/home/salgado/src/async-service/async_service/trio.py", line 65, in _handle_cancelled
    for task in iter_dag(self._service_task_dag):
  File "/home/salgado/src/async-service/async_service/_utils.py", line 56, in iter_dag
    for item in dag.keys():
RuntimeError: dictionary changed size during iteration
```

**Comment by gsalgado** [comment]
Sometimes I do get the correct exceptions, so yeah, definitely a race condition somewhere:
```
13:34:28 DEBUG: bonding completed successfully with <Node(0x15ac30@52.176.7.10)>
Traceback (most recent call last):
  File "scripts/discovery.py", line 67, in <module>
    trio.run(main)
  File "/home/salgado/virtualenvs/trinity-trio/lib/python3.7/site-packages/trio/_core/_run.py", line 1804, in run
    raise runner.main_task_outcome.error
  File "scripts/discovery.py", line 63, in main
    await service.manager.wait_finished()
  File "/home/salgado/virtualenvs/trinity-trio/lib/python3.7/site-packages/async_generator/_util.py", line 42, in __aexit__
    await self._agen.asend(None)
  File "/home/salgado/src/async-service/async_service/trio.py", line 413, in background_trio_service
    await manager.stop()
  File "/home/salgado/virtualenvs/trinity-trio/lib/python3.7/site-packages/trio/_core/_run.py", line 730, in __aexit__
    raise combined_error_from_nursery
  File "/home/salgado/src/async-service/async_service/trio.py", line 156, in run
    for _, exc_value, exc_tb in self._errors
trio.MultiError: AssertionError(), LifecycleError('Tasks may not be scheduled if the service is not running')

Details of embedded exception 1:

  Traceback (most recent call last):
    File "/home/salgado/src/async-service/async_service/trio.py", line 241, in _run_and_manage_task
      await async_fn(*args)
    File "/home/salgado/src/trinity/p2p/discovery.py", line 223, in fetch_enrs
      raise AssertionError()
  AssertionError

Details of embedded exception 2:

  Traceback (most recent call last):
    File "/home/salgado/src/async-service/async_service/trio.py", line 241, in _run_and_manage_task
      await async_fn(*args)
    File "/home/salgado/src/trinity/p2p/discovery.py", line 238, in _ensure_enr
      await self.get_enr(node)
    File "/home/salgado/src/trinity/p2p/discovery.py", line 356, in get_enr
      enr = await self.request_enr(remote)
    File "/home/salgado/src/trinity/p2p/discovery.py", line 312, in request_enr
      token = self.send_enr_request(remote)
    File "/home/salgado/src/trinity/p2p/discovery.py", line 670, in send_enr_request
      message = self.send(node, CMD_ENR_REQUEST, [_get_msg_expiration()])
    File "/home/salgado/src/trinity/p2p/discovery.py", line 514, in send
      self.socket.sendto, message, (node.address.ip, node.address.udp_port))
    File "/home/salgado/src/async-service/async_service/trio.py", line 274, in run_task
      "Tasks may not be scheduled if the service is not running"
  async_service.exceptions.LifecycleError: Tasks may not be scheduled if the service is not running
```

**Comment by pipermerriam** [comment]
I'm thinking that we may need to use a lock to manage access to the task dag.  My reasoning for this is that we want to ensure that once we start cancelling things we don't schedule anything new (modify the dag).  If we protect it with a lock then we should be able to fix this because **either** the lock will be held by the cleanup code in which case all attempts to schedule new stuff should hang attempting to acquire the lock, or the lock will be held by the code that is starting the task at which point the cleanup code will wait until the task is fully scheduled before doing cleanup.

I think this should fix both scenarios.

**Comment by gsalgado** [comment]
Makes sense, that should probably do the trick indeed

**Comment by pipermerriam** [comment]
I think this should be fixed in `0.1.0-alpha.5`.  If it shows back up please re-open.
